### PR TITLE
Optimize the ADIOS conversion tool

### PR DESCRIPTION
**Pull request by dqwu**
This patch is provided by Tahsin Kurc, which further improves the
performance of the ADIOS conversion tool.

Major changes:
Use SUBSET rearranger by default.
Merge start/count/buffer to reduce costly I/O seeks.
Reorganize order to process variables to reduce decomposition
allocations.

Note the updated conversion tool will not work with the previously
generated BP files,  because it will expect merged start/count/buffer.

### CODE REVIEW

File: tools/adios2pio-nm/adios2pio-nm.cxx [hunk 1]
```diff
@@ -13,7 +13,7 @@ static void init_user_options(spio_tool_utils::ArgParser &ap)
       .add_opt("idir", "Directory containing data output from PIO (in ADIOS format)")
       .add_opt("nc-file", "output file name after conversion")
       .add_opt("pio-format", "output PIO_IO_TYPE. Supported parameters: \"pnetcdf\",  \"netcdf\",  \"netcdf4c\",  \"netcdf4p\"")
-      .add_opt("reduce-memory-usage", "Reduce memory usage (execution time will likely increase)")
+      .add_opt("rearr", "PIO rearranger. Supported parameters: \"subset\", \"box\". Default \"subset\".")
```
Comment (jayeshkrishna) [inline tools/adios2pio-nm/adios2pio-nm.cxx:16]: Is the "reduce memory" option no longer required?

**Review by jayeshkrishna — APPROVED** [review]

File: tools/adios2pio-nm/adios2pio-nm.cxx (hunk 1 above)
Comment (dqwu) [reply tools/adios2pio-nm/adios2pio-nm.cxx:16]: I think so. We can confirm it with @tkurc ,though.

**Comment by dqwu** [comment]
This fix has been tested on Cori with eam.r BP files output by
E3SM ne4 F case (16 BP files).

The conversion tool is built with CMAKE_BUILD_TYPE=Release
to turn on compiler optimizations.

This fix also reduces the total size of BP files output by E3SM
ne4 F case runs:
eam.r BP files size before fix: 567 MB
eam.r BP files size after fix: 411 MB

[Cori Haswell node, run with 16 MPI tasks]
conversion time before fix, BOX rearranger: 35 sec
conversion time before fix, SUBSET rearranger: 34 sec

conversion time after fix, BOX rearranger: 20 sec
conversion time after fix, SUBSET rearranger: 23 sec

[Cori KNL node, run with 16 MPI tasks]
conversion time before fix, BOX rearranger: 124 sec
conversion time before fix, SUBSET rearranger: 129 sec

conversion time after fix, BOX rearranger: 48 sec
conversion time after fix, SUBSET rearranger: 54 sec

File: tools/adios2pio-nm/adios2pio-nm.cxx (hunk 1 above)
Comment (tkurc) [reply tools/adios2pio-nm/adios2pio-nm.cxx:16]: No need for reduce memory option.

**Review by tkurc — APPROVED** [review]

**Comment by dqwu** [comment]
This fix has also been tested on Cori KNL nodes with eam.r BP files
output by E3SM scream ne1024 F case (1536 BP files).

Each BP file has a smaller size (about 10 MB less) with this fix:
Single eam.r BP file size before fix: 1,996,862,704 bytes
Total (1536) BP files size before fix: 2.789584972 TB

Single eam.r BP file size after fix: 1,986,537,854 bytes
Total (1536) BP files size after fix: 2.775161323 TB

[48 Cori KNL nodes, 1536 MPI tasks]
conversion time before fix, BOX rearranger: 29 min 7 sec
conversion time before fix, SUBSET rearranger: 19 min 20 sec

conversion time after fix, BOX rearranger: 18 min 10 sec
conversion time after fix, SUBSET rearranger: 17 min 44 sec

**Comment by dqwu** [comment]
By freeing a decomposition that is no longer being used as early
as possible, this fix allows users to run the conversion tool with
fewer KNL nodes on Cori. With available memory reduced, the
conversion time might be longer, though.

[24 Cori KNL nodes, 1536 MPI tasks]
conversion time before fix, BOX rearranger: Out Of Memory
conversion time before fix, SUBSET rearranger: Out Of Memory

conversion time after fix, BOX rearranger: 39 min 14 sec
conversion time after fix, SUBSET rearranger: 28 min 21 sec

**Comment by dqwu** [comment]
This fix has also been tested with E3SM test suite for correctness.

**Pull request merged**